tubular/scripts/prune_modulestore.py

```
@@ -0,0 +1,559 @@
+                # end of the tree (original version)
+                break
+
+            version_tree.append(str(structure_doc['_id']))
+
+    return version_tree
+
+
+def get_structures_to_delete(active_versions, structures=None, version_retention=2, relink_structures=False):
+
+    """
+    Generate a list of structures that meet the conditions for pruning and associated visualization
+    """
+
+    # initialize key variables
+    version_trees = []
+    versions_to_retain = []
+    versions_to_remove = []
+    counter = 0
+
+    # iterate the active versions and build the associated version tree
+    for active_version in active_versions:
+
+        counter += 1
+
+        for target_key in TARGET_ACTIVE_VERSIONS_KEYS:
+
+            if target_key in active_version['versions']:
+
+                # print(active_version)
+                version_tree = build_activeversion_tree(active_version['versions'][target_key], structures)
+
+                # only add the tree if it has 1+ element
+                tree_length = len(version_tree)
+                if tree_length > 0:
+
+                    status_message = "Processing Active Version {0} | {3}: {1} version with a {2}-version tree"
+
+                    LOG.debug(status_message.format(
+                        counter,
+                        target_key,
+                        len(version_tree),
+                        str(active_version['_id'])))
+
+                    # if the tree exceeds the minimum number of elements,
+                    # identify tree elements that should be removed
+                    if tree_length > version_retention:
+
+                        # track the required version: first & last
+                        versions_to_retain.extend(version_tree[:2])
+
+                        # if relinking is not required, it is ok to remove the original version
+                        if not relink_structures:
+                            versions_to_retain.append(version_tree[-1])
```

I don't think we ever want to remove the original version, even if it's a meaningless empty course. I thought that the process of relinking meant that we would be modifying the "previous" pointer of the oldest preserved structure to point to "original". The original is used to determine things like common ancestry, and due to the way things are cached, we can get weird results if we change the original and do updates.

For instance, we might still have a cached current structure doc that points to original_1. If the pruning process deletes original_1 and updates the structure docs in MongoDB to point to original_2 as their new "original", then looking for ancestors (e.g. split's `find_ancestor_structures()`) against the cached original will silently fail in a hard to debug way.